File: haskell/life.hs
-- Game of Life, toroidal 64x64 — purely functional form: each generation is
-- a fresh immutable unboxed array built from the previous one (the
-- tabulate-equivalent). Same seed hash, rule, and wrap as the other
-- variants.
{-# LANGUAGE BangPatterns #-}

module Main where

import Data.Array.Unboxed
import Data.Int (Int32)
import System.Exit (exitFailure)

n :: Int
n = 64

cells :: Int
cells = n * n

wrap :: Int -> Int
wrap x = (x + 64) `rem` 64

step :: UArray Int Int32 -> UArray Int Int32
step g = listArray (0, cells - 1) [cell k | k <- [0 .. cells - 1]]
  where
    cell k =
      let i = k `quot` 64
          j = k `rem` 64
          im = wrap (i - 1)
          ip = wrap (i + 1)
          jm = wrap (j - 1)
          jp = wrap (j + 1)
          nb =
            g ! (im * 64 + jm) + g ! (im * 64 + j) + g ! (im * 64 + jp)
              + g ! (i * 64 + jm) + g ! (i * 64 + jp)
              + g ! (ip * 64 + jm) + g ! (ip * 64 + j) + g ! (ip * 64 + jp)
          alive = g ! (i * 64 + j)
       in if alive == 1
            then (if nb == 2 || nb == 3 then 1 else 0)
            else (if nb == 3 then 1 else 0)

run :: Int -> UArray Int Int32 -> UArray Int Int32
run 0 !g = g
run k !g = run (k - 1) (step g)

seedGrid :: UArray Int Int32
seedGrid = listArray (0, cells - 1) [cell k | k <- [0 .. cells - 1]]
  where
    cell k =
      let i = k `quot` 64
          j = k `rem` 64
       in if (i * 2654435761 + j * 40503 + i * j * 2246822519) `rem` 97 < 33
            then 1
            else 0

population :: UArray Int Int32 -> Int
population g = sum (map fromIntegral (elems g))

main :: IO ()
main = do
  let pop = population (run 50000 seedGrid)
  if pop /= 115
    then putStrLn ("FAIL: expected 115, got " ++ show pop) >> exitFailure
    else pure ()

File: ocaml/life.ml
let side = 64
let cells = side * side
let wrap x = (x + side) mod side

let step grid =
  Array.init cells (fun k ->
      let i = k / side and j = k mod side in
      let im = wrap (i - 1) and ip = wrap (i + 1) in
      let jm = wrap (j - 1) and jp = wrap (j + 1) in
      let neighbours =
        grid.((im * side) + jm) + grid.((im * side) + j)
        + grid.((im * side) + jp) + grid.((i * side) + jm)
        + grid.((i * side) + jp) + grid.((ip * side) + jm)
        + grid.((ip * side) + j) + grid.((ip * side) + jp)
      in
      let alive = grid.((i * side) + j) in
      if alive = 1 then if neighbours = 2 || neighbours = 3 then 1 else 0
      else if neighbours = 3 then 1
      else 0)

let rec run grid generations =
  if generations = 0 then grid else run (step grid) (generations - 1)

let seed =
  Array.init cells (fun k ->
      let i = k / side and j = k mod side in
      if ((i * 2_654_435_761) + (j * 40_503) + (i * j * 2_246_822_519)) mod 97 < 33
      then 1
      else 0)

let () =
  let population = Array.fold_left ( + ) 0 (run seed 50_000) in
  if population <> 115 then (
    Printf.eprintf "FAIL: expected 115, got %d\n" population;
    exit 1)
